File: Day18.java
// Runtime : 4ms
class Solution {
    public String countAndSay(int n) {
	    String res ="1";
	    if(n==1){
	        return res;
	    }else{
	        for(int i=2;i<=n;i++){
	            String str =countsay(res);
	            res = str;
	        }
            return res;
	    }
    }
    public String countsay(String code){
        StringBuilder sb =new StringBuilder();
        int count=1;
        for(int i=1;i<code.length();i++){
            if(code.charAt(i)==code.charAt(i-1)){
                count++;
            }else{
                sb.append(count);
                sb.append(code.charAt(i-1));
                count=1;
            }
        }
        sb.append(count);
        sb.append(code.charAt(code.length()-1));
        return sb.toString();
    }
}

File: Day18.cpp
//-----------------------Recursion--------------------------------

class Solution {
public:
    string countAndSay(int n) {
        //base case
        if(n == 1) return "1";
        //get the previous, which will help us in finding current.
        string prev = countAndSay(n-1);
        //current answer
        string ans;
        //conversion process starts.
        int len = prev.length();
        //traverse the string
        for(int i=0; i<len;){
            //count frequency 
            int j=i, cnt = 0;
            while(j< len){
                if(prev[i] == prev[j]){
                    cnt++;
                    j++;
                }else{
                    break;
                }
            }
            //add it in answer
            ans.push_back(cnt + '0');
            ans.push_back(prev[i]);
            i = j;
        }
        return ans;
    }
};


//--------------------------Tabulation--------------------------------

class Solution {
public:
    string countAndSay(int n) {
        vector<string>arr(n);
        arr[0] = "1";

        for(int i=1; i<n; i++){
            string prev = arr[i-1], curr;
            int st = 0, cnt = 1;
            for(int j = 1; j<prev.length(); j++){
                if(prev[j] == prev[st]) cnt++;
                else{
                    curr.push_back(cnt+ '0');
                    curr.push_back(prev[st]);
                    cnt = 1; st = j;
                }
            }
            curr.push_back(cnt+ '0');
            curr.push_back(prev[st]);

            arr[i] = curr;
        }

        return arr[n-1];
    }
};


//------------------------Space Optimized------------------------------


class Solution {
public:
    string countAndSay(int n) {
        string prev = "1";

        for(int i=2; i<=n; i++){
            string curr;
            int st = 0, cnt = 1;
            for(int j = 1; j<prev.length(); j++){
                if(prev[j] == prev[st]) cnt++;
                else{
                    curr.push_back(cnt+ '0');
                    curr.push_back(prev[st]);
                    cnt = 1; st = j;
                }
            }
            curr.push_back(cnt+ '0');
            curr.push_back(prev[st]);

            prev = curr;
        }

        return prev;
    }
};